mrjob/emr.py

```diff
@@ -2603,45 +2619,52 @@ def _find_cluster(self, num_steps=1):
         compute units. Break ties by choosing cluster with longest idle time.
         Return ``None`` if no suitable clusters exist.
         """
-        exclude = set()
+        valid_clusters = {}
+        invalid_clusters = set()
+        locked_clusters = set()
+
         max_wait_time = self._opts['pool_wait_minutes']
+        bypass_pool_wait = self._opts['bypass_pool_wait']
         now = datetime.now()
         end_time = now + timedelta(minutes=max_wait_time)
-        time_sleep = timedelta(seconds=_POOLING_SLEEP_INTERVAL)
 
         log.info('Attempting to find an available cluster...')
         while now <= end_time:
             cluster_info_list = self._usable_clusters(
-                exclude=exclude,
-                num_steps=num_steps)
+                valid_clusters, invalid_clusters,
+                locked_clusters, num_steps)
```

Comment:
I don't know if they use asserts anywhere in this code but I would add 
`assert len(valid_clusters & invalid_clusters) == 0